aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -488,6 +489,23 @@ Tensor index_fill(const Tensor & self, int64_t dim, const Tensor & index, const
   return self.clone(at::MemoryFormat::Preserve).index_fill_(dim, index, source);
 }
 
+Tensor & gather_out_cpu(Tensor & result, const Tensor & self, int64_t dim, const Tensor & index, bool sparse_grad) {
+  if (result.defined()) {
+    result.resize_as_(index);
+  }
+  else {
+    result = at::empty(index.sizes(), self.options());
+  }
+
+  gather_stub(result.device().type(), result, self, dim, index);
+  return result;
+}
```

Comment:
@ngimel , what do you think about modifying `gather` such that `self.gather(dim, index)` can actually modify `self`, or return a tensor `res` such that `res.shape = self.shape`, where non-indexed elements (indices outside of `index.shape`) are set to zero? This way it will be possible to fix backward formula for `scatter_add`, which is incorrect as it is right now.

Reply:
Returning an extra tensor, especially when user did not ask for it, does not seem like a great idea. However, having a non-exposed internal API that would do that, and call that API for `scatter_add` gradient is ok. It can share a lot of implementation with regular gather, so that both this new function and regular `gather` would call into a common `_impl` function. The only difficulty here is that for double backward you'd also have to define a gradient for this new function.

Reply:
No extra tensors! Just extend the parameter list by `result_shape` which defaults to `index.shape` if not provided.

Reply:
Sorry, did not pay enough attention. Still, to me it looks like not enough motivation to change behavior of `gather`, I'd prefer it to be `gather` as it currently is, not currently exposed `gather_with_shape` to use in the backward of scatter, both calling into `gather_impl` that does the right thing. Do you find this plan unworkable?

Reply:
What do you mean by non-exposed? Like it does not become a part of a Python API? How do I do that? Is it the underscore methods?

Reply:
Yep, the underscore method.